src/backend/cuda/kernel/moments.hpp

```diff
@@ -0,0 +1,138 @@
+/*******************************************************
+ * Copyright (c) 2014, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <af/defines.h>
+#include <backend.hpp>
+#include <Param.hpp>
+#include <dispatch.hpp>
+#include <math.hpp>
+#include <err_cuda.hpp>
+#include <debug_cuda.hpp>
+
+namespace cuda
+{
+namespace kernel
+{
+
+    // Kernel Launch Config Values
+    static const int THREADS = 128;
+
+    // Moment functions
+    template<typename T>
+    __device__ inline static
+    T moments_m00(const dim_t idx, const dim_t idy, const dim_t idz, const dim_t idw,
+                       CParam<T> in, const bool pBatch)
+    {
+        dim_t mId = idy * in.strides[1] + idx;
+        if(pBatch) {
+            mId += idw * in.strides[3] + idz * in.strides[2];
+        }
+
+        return in.ptr[mId];
+    }
+
+    template<typename T>
+    __device__ inline static
+    T moments_m01(const dim_t idx, const dim_t idy, const dim_t idz, const dim_t idw,
+                       CParam<T> in, const bool pBatch)
+    {
+        dim_t mId = idy * in.strides[1] + idx;
+        if(pBatch) {
+            mId += idw * in.strides[3] + idz * in.strides[2];
+        }
+
+        return idx * in.ptr[mId];
+    }
+
+    template<typename T>
+    __device__ inline static
+    T moments_m10(const dim_t idx, const dim_t idy, const dim_t idz, const dim_t idw,
+                       CParam<T> in, const bool pBatch)
+    {
+        dim_t mId = idy * in.strides[1] + idx;
+        if(pBatch) {
+            mId += idw * in.strides[3] + idz * in.strides[2];
+        }
+
+        return idy * in.ptr[mId];
+    }
+
+    template<typename T>
+    __device__ inline static
+    T moments_m11(const dim_t idx, const dim_t idy, const dim_t idz, const dim_t idw,
+                       CParam<T> in, const bool pBatch)
+    {
+        dim_t mId = idy * in.strides[1] + idx;
+        if(pBatch) {
+            mId += idw * in.strides[3] + idz * in.strides[2];
+        }
+
+        return idx * idy * in.ptr[mId];
+    }
+
+    template<typename T, af_moment moment>
+    __global__
+    void moments_kernel(CParam<float> out, CParam<T> in,
+                  const dim_t blocksMatX, const bool pBatch)
+    {
+        const dim_t idw = blockIdx.y / in.dims[2];
+        const dim_t idz = blockIdx.y - idw * in.dims[2];
```

Comment:
you can check for `pBatch` here and offset the last two dimensions before calling the functors.